### ISSUE TITLE
NullPointerException when overriding copts and linkopts to None in transition

### ISSUE THREAD

**Issue by stefanbucur** [bazelbuild/bazel#12559]
### Description of the problem / feature request:

When overriding the `copts` and `linkopts` build configuration with `None` values in a transition, Bazel crashes with a NullPointerException.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```py
def _my_transition_impl(settings, attr):
    return {
        "//command_line_option:copt": None,
        "//command_line_option:linkopt": None,
    }

my_transition = transition(
    implementation = _my_transition_impl,
    inputs = [],
    outputs = [
        "//command_line_option:copt",
        "//command_line_option:linkopt",
    ],
)
```

### What operating system are you running Bazel on?

Debian-based Linux.

### What's the output of `bazel info release`?

`release 3.7.1`

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

```
git@github.com:bazelbuild/rules_fuzzing
f782cd326f15a1182a3c293cf77f3a21f0b24ed0
b0e592b75de5680599ee81ccec6b073e93a21058
```

###  Have you found anything relevant by searching the web?

No.

### Any other information, logs, or outputs that you want to share?

The exception stack trace:

```
Internal error thrown during build. Printing stack trace: java.lang.RuntimeException: Unrecoverable error while evaluating node 'BuildConfigurationValue.Key[f380618cb25b3c02e111c0b5e52b926621d76dbf320be25008319e76d414f029]' (requested by nodes 'ConfiguredTargetKey{label=//examples:re2_fuzz_test_instrum, config=BuildConfigurationValue.Key[8f956e0a5a332769538bb1785bbb03d792f84521f14f8711180a17d7de3e07e8]}')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:513)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:398)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.NullPointerException
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
        at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
        at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
        at com.google.devtools.build.lib.skyframe.BuildConfigurationFunction.getConfigurationFragments(BuildConfigurationFunction.java:127)
        at com.google.devtools.build.lib.skyframe.BuildConfigurationFunction.compute(BuildConfigurationFunction.java:84)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:438)
        ... 7 more
Caused by: java.lang.NullPointerException
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:878)
        at com.google.common.collect.ImmutableList$Builder.addAll(ImmutableList.java:823)
        at com.google.devtools.build.lib.rules.cpp.CppConfiguration.create(CppConfiguration.java:182)
        at com.google.devtools.build.lib.rules.cpp.CppConfigurationLoader.create(CppConfigurationLoader.java:49)
        at com.google.devtools.build.lib.rules.cpp.CppConfigurationLoader.create(CppConfigurationLoader.java:27)
        at com.google.devtools.build.lib.skyframe.BuildConfigurationFunction.makeFragment(BuildConfigurationFunction.java:158)
        at com.google.devtools.build.lib.skyframe.BuildConfigurationFunction.access$000(BuildConfigurationFunction.java:45)
        at com.google.devtools.build.lib.skyframe.BuildConfigurationFunction$2.load(BuildConfigurationFunction.java:59)
        at com.google.devtools.build.lib.skyframe.BuildConfigurationFunction$2.load(BuildConfigurationFunction.java:56)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
        ... 13 more
```

**Comment by aiuto** [comment]
@stefanbucur Does using "" instead of None work for you?
This is a bug either way. I'm just trying to understand severity so we can prioritize.

**Comment by stefanbucur** [comment]
> @stefanbucur Does using "" instead of None work for you?

Using `""` shows a different error:

```
ERROR: [...]/bazel-rules-fuzzing/fuzzing/private/instrument.bzl:33:5: IllegalArgumentError for option 'copt': Can not set java.util.List field com.google.devtools.build.lib.rules.cpp.CppOptions.coptList to java.lang.String
ERROR: Errors encountered while applying Starlark transition
```

However, it does work when I use a list type instead (`[]`), presumably because `--copt` are list-typed command-line options.

**Comment by aiuto** [comment]
My bad. I should have said [] rather than "".
Thanks for checking. With an easy workaround, we know better how to prioritize it.

**Comment by juliexxia** [comment]
(Context: I'm moving on from Bazel on friday and spending some time triaging all bugs assigned to me before unassigning myself)

Both //command_line_option:copt and //command_line_option:linkopt should use `[]` instead of None to indicate an empty value since they are stored as [List types](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java;l=321?q=CppOptions.java&ss=bazel). Yes, we should change the code such that it doesn't NPE though.

**Comment by tobiade** [comment]
Hey, newbie here - I've seen this has been marked as a good first issue, so I'm happy to pick this up. If you can let me know what the expectation is for the resolution of this issue, that'll be great e.g a more useful exception message rather than an NPE? Or some sensible defaults?

In addition, if you can point me to an area of code where I can start poking around that'll be appreciated (I'll also dig through the stack trace).

Thanks!

**Comment by stefanbucur** [comment]
My $0.02 (not on the Bazel team, so deferring the decision on the semantics here to them): I think it would be nice if Bazel would show a compilation error when using the `None` value for any configuration value in the transition output. It currently hard-crashes with a Java exception.

**Comment by gregestren** [comment]
Hi @tobiade -  I just saw this thread. Happy you're interested in contributing! First and foremost, have you gone through the basics of simply building Bazel? Like make a trivial change (to a string or something) and verify you can build and run a custom Bazel with it?

Happy to offer more input on this particular issue from that point. I'll have to review the thread to remember details, but can definitely better scope the issue.

**Comment by tobiade** [comment]
Thanks! @gregestren - I've been able to successfully build [Bazel from source](https://docs.bazel.build/versions/master/install-compile-source.html), but been having a bit of a nightmare trying to get Bazel set up in Intellij on Windows using th Intellij plugin. I did eventually find out that the plugin isn't supported on Windows, so I'm looking for a workaround now e.g WSL2 + IntelliJ.

**Comment by gregestren** [comment]
Sorry about the IntelliJ problems. I haven't had any issues on Linux, so hopefully that approach works.

The core issue, I believe, is there are different kinds of data typing in Bazel. When you set a value to `None`, you're setting it to a type in the Starlark language. Transitions ultimately set that to pure Java types. For example, `copt` is [ultimately](https://github.com/bazelbuild/bazel/blob/2d28044316bf1ea9935f85e9a6876062abce169f/src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java#L278) a Java `List<String>`.

I think the specific code doing this translation is at https://github.com/bazelbuild/bazel/blob/98d376faeb206f14838156ce4cb305ddbfce08fa/src/main/java/com/google/devtools/build/lib/analysis/starlark/FunctionTransitionUtil.java#L297-L299 So this may be a matter of examining the option's type, and if it's a list type, change it to `[]` instead of `null`?

An alternative approach could be to ban setting values to `None`, so examples like the above simply report "This is not a valid assignment`.  With that approach, maybe the same  would trigger an error message instead of trying to change the conversion. 

But for non-list types, setting to `None` makes more sense. So maybe the first approach is more consistent.

Is that enough to work with?

**Comment by tobiade** [comment]
Hey! Sorry about the delay - thanks, that makes sense. First  approach sounds good to me - consistent, and probably safe enough to infer that the user actually means `[]`. I'm looking into it now.

**Comment by gregestren** [comment]
@stefanbucur what was your expected value with `None`. "Empty" or "default"? https://github.com/bazelbuild/bazel/pull/13286#discussion_r605309466 discusses the idea of it meaning default. 

I also acknowledge in https://github.com/bazelbuild/bazel/issues/12559#issuecomment-800353573 you suggested simply disallowing `None`.

**Comment by stefanbucur** [comment]
(Sorry for the delay, for some reason I missed the notification.)

Yeah, I was suggesting to disallow `None`, because it can be confusing at the use site. Without the reader being familiar with the semantics, I think it can ambiguously be interpreted as either one of these.

By the way, if "default" here means to leave the configuration value unchanged, then this begs the question of whether the configuration should also be listed as an input in the transition, since in a way it implicitly adopts the input value.

**Comment by gregestren** [comment]
By "default" I mean the flag's implicit value at the command line if you never explicitly set it.

As a random example, consider [`--crosstool_top`](https://github.com/bazelbuild/bazel/blob/4571417a35fd4f86316417edebab9c083a503244/src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java#L116):

    $ bazel build //foo          # --crosstool_top = @bazel_tools/tools/cpp:toolchain (default)
    $ bazel build //foo --crosstool_top=//bar       # --crosstool_top=//bar
    $ bazel build //foo --crosstool_top=""          # --crosstool_top is empty

So a transition could change a flag's value to its default, which could be different from its input value. In other words, with:

    $ bazel build //foo --crosstool_top=//bar

a transition that sets

    --crosstool_top=@bazel_tools/tools/cpp:toolchain

sets `--crosstool_top` back to its default.

**Comment by tobiade** [comment]
This has now been fixed by #13286. Thanks @gregestren @stefanbucur!
